cmd/kubeadm/app/util/version.go

```diff
@@ -72,11 +76,19 @@ func KubernetesReleaseVersion(version string) (string, error) {
 		return ver, nil
 	}
 
+	// kubeReleaseLabelRegex matches labels such as: latest, latest-1, latest-1.10
 	if kubeReleaseLabelRegex.MatchString(versionLabel) {
 		url := fmt.Sprintf("%s/%s.txt", bucketURL, versionLabel)
 		body, err := fetchFromURL(url, getReleaseVersionTimeout)
 		if err != nil {
-			return "", err
+			// Handle air-gapped environments by falling back to the client version.
+			glog.Infof("could not fetch a Kubernetes version from the internet: %v", err)
```

Comment:
That fallback is not good. We should go to this fallback only in cases where we have technical issues to reach download server (timeouts, no-route-to-host, DNS failure, etc). If it was able to connect to the server, but requested label is not found (HTTP error 404), then this error about incorrect label should be propagated up.

Reply:
i agree that 404 should be handled here.

one problem, in an air-gapped scenario passing "foo" would fail fetching the version from the server with an error different from 404, this would then fallback to the client version but we don't validate a client version locally in any way and "foo" would end up being a valid version and this is bogus.

thus i think the local version should be handled with this regexp when the fallback to the client version happens:
`^(latest|stable)(-([0-9]+)(\.([0-9]+))?)?$`

from the previous comment about the regexp:
> We don't want to modify that piece of code every time when somebody creates new label on release server.

i don't see this being a big problem. if the servers change or labels change we would have to make changes regardless.

Reply:
Wait. There is actually another issue here in your patch :)
So, for cases with fallback (result of `kubeadmVersion()` function from your code), it should be assigned to the `body` variable and let it re-validate below via recursive execution of this function. It shouldn't return immediately.

Regarding airgap issues: with 404 it is clear answer, return error. In all other scenarios (timeout, HTTP error code 5xx, etc) - fallback. We just need to add test cases for those HTTP replies in the network test piece for this code:
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/util/version_test.go#L115